### ISSUE TITLE
1.21 HA Cluster - kubelite constantly crashing - dqlite issues?

### ISSUE THREAD

**Issue by MichaelGooden** [canonical/microk8s#2285]
I am facing issues using the HA feature. The cluster is completely unstable and kubelite crashes constantly on all nodes.

[inspection-report-20210524_202530.tar.gz](https://github.com/ubuntu/microk8s/files/6534335/inspection-report-20210524_202530.tar.gz)
[inspection-report-20210524_202412.tar.gz](https://github.com/ubuntu/microk8s/files/6534336/inspection-report-20210524_202412.tar.gz)
[inspection-report-20210524_202223.tar.gz](https://github.com/ubuntu/microk8s/files/6534337/inspection-report-20210524_202223.tar.gz)

Any ideas?

**Comment by ktsakalozos** [comment]
Could you please benchmark the disk performance?

Looking at the logs there are a lot of messages like this:
```
May 24 20:22:40 k8s05 microk8s.daemon-kubelite[22269]: Trace[950610081]: ---"Transaction committed" 9257ms (20:22:00.905)
```
This indicates the I/O disk performance causes the datastore to be very slow. 

Have a look at the `Raspberry Pi and systems with low disk performance` case of the `Common issues` [1] docs section.

[1] https://microk8s.io/docs/troubleshooting#heading--common-issues

**Comment by MichaelGooden** [comment]
Very strange, these nodes are VMs backed by SSD storage:

```
k8s04

1+0 records in
1+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 4.16401 s, 258 MB/s

k8s05

1+0 records in
1+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 3.16778 s, 339 MB/s

k8s06

1+0 records in
1+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 3.09161 s, 347 MB/s
```

**Comment by MichaelGooden** [comment]
hdparm:
```
k8s04

/dev/xvda1:
 Timing cached reads:   7218 MB in  1.96 seconds = 3683.66 MB/sec
 Timing buffered disk reads: 394 MB in  3.01 seconds = 131.00 MB/sec

k8s05

/dev/xvda1:
 Timing cached reads:   8806 MB in  1.97 seconds = 4458.90 MB/sec
 Timing buffered disk reads: 472 MB in  3.01 seconds = 156.83 MB/sec

k8s06

/dev/xvda1:
 Timing cached reads:   9246 MB in  1.97 seconds = 4693.34 MB/sec
 Timing buffered disk reads: 454 MB in  3.01 seconds = 151.03 MB/sec
```

**Comment by ktsakalozos** [comment]
What about the network between the xen VMs?

**Comment by MichaelGooden** [comment]
Network appears fine, VM to VM pings are ~0.5ms.

Plenty of throughput available. The NICs that carry the VM traffic are not congested (<5% utilisation)

```
------------------------------------------------------------
Server listening on TCP port 5001
TCP window size:  128 KByte (default)
------------------------------------------------------------
[  4] local k8s06 port 5001 connected with k8s05 port 48034
[ ID] Interval       Transfer     Bandwidth
[  4]  0.0-10.0 sec  1.09 GBytes   930 Mbits/sec
[  4] local k8s06 port 5001 connected with k8s04 port 34964
[  4]  0.0-10.0 sec  1.07 GBytes   917 Mbits/sec
^Croot@k8s06:~# iperf -c k8s04
------------------------------------------------------------
Client connecting to k8s04, TCP port 5001
TCP window size: 85.0 KByte (default)
------------------------------------------------------------
[  3] local k8s06 port 48110 connected with k8s04 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-10.0 sec  1.05 GBytes   905 Mbits/sec
root@k8s06:~# iperf -c k8s05
------------------------------------------------------------
Client connecting to k8s05, TCP port 5001
TCP window size: 85.0 KByte (default)
------------------------------------------------------------
[  3] local k8s06 port 47500 connected with k8s05 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-10.0 sec  1.06 GBytes   914 Mbits/sec
```

**Comment by ktsakalozos** [comment]
Do you see this behavior on a single node cluster as well?

Is this issue present in the 1.20 channel?

@MathieuBordere any ideas what may be causing this?

**Comment by ktsakalozos** [comment]
@MichaelGooden could you share the output of `sudo ls -l  /var/snap/microk8s/current/var/kubernetes/backend/` ?

**Comment by MathieuBordere** [comment]
> Do you see this behavior on a single node cluster as well?
> 
> Is this issue present in the 1.20 channel?
> 
> @MathieuBordere any ideas what may be causing this?

Not at the moment, waiting on the contents of the database folder for some more info.

**Comment by MichaelGooden** [comment]
@ktsakalozos I experienced this on 1.20 as well. I switched that cluster to non-HA and it has been working since. I created a new cluster on new VMs on 1.21, this is the cluster we are currently debugging.

The behaviour only became apparent after a few days, initially the cluster performed as expected. This was the same behaviour observed in the previous cluster on 1.20.

I have not run a single node HA-enabled cluster for extended durations, I will spin one up now and let it idle.

I have attempted the HA recovery process a few times, it did not make a difference.

Please see the database directory listings as requested:
[k8s06.txt](https://github.com/ubuntu/microk8s/files/6537742/k8s06.txt)
[k8s05.txt](https://github.com/ubuntu/microk8s/files/6537743/k8s05.txt)
[k8s04.txt](https://github.com/ubuntu/microk8s/files/6537744/k8s04.txt)

**Comment by MathieuBordere** [comment]
Strange, there is no `snapshot-451-11678894-46572435.meta` file in any of the folders, there should be one for the snapshot to be correctly restored.

**Comment by MichaelGooden** [comment]
GH mangled the file upload...

k8s04:
```
6.6M -rw-rw---- 1 root microk8s 6.6M May 25 06:24 /var/snap/microk8s/current/var/kubernetes/backend/0000000011671634-0000000011671889
5.4M -rw-rw---- 1 root microk8s 5.4M May 25 06:31 /var/snap/microk8s/current/var/kubernetes/backend/0000000011671890-0000000011672104
3.6M -rw-rw---- 1 root microk8s 3.6M May 25 06:35 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672105-0000000011672245
6.7M -rw-rw---- 1 root microk8s 6.7M May 25 06:41 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672246-0000000011672503
5.6M -rw-rw---- 1 root microk8s 5.6M May 25 06:46 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672504-0000000011672727
3.0M -rw-rw---- 1 root microk8s 2.9M May 25 06:49 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672728-0000000011672845
4.7M -rw-rw---- 1 root microk8s 4.7M May 25 06:53 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672846-0000000011673026
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 07:01 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673027-0000000011673344
3.4M -rw-rw---- 1 root microk8s 3.4M May 25 07:05 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673345-0000000011673476
2.9M -rw-rw---- 1 root microk8s 2.9M May 25 07:08 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673477-0000000011673586
5.6M -rw-rw---- 1 root microk8s 5.6M May 25 07:14 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673587-0000000011673807
2.2M -rw-rw---- 1 root microk8s 2.2M May 25 07:17 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673808-0000000011673901
4.5M -rw-rw---- 1 root microk8s 4.5M May 25 07:22 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673902-0000000011674085
4.7M -rw-rw---- 1 root microk8s 4.7M May 25 07:27 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674086-0000000011674276
5.2M -rw-rw---- 1 root microk8s 5.2M May 25 07:32 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674277-0000000011674478
3.6M -rw-rw---- 1 root microk8s 3.6M May 25 07:37 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674479-0000000011674615
4.7M -rw-rw---- 1 root microk8s 4.7M May 25 07:43 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674616-0000000011674814
3.8M -rw-rw---- 1 root microk8s 3.8M May 25 07:48 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674815-0000000011674965
2.2M -rw-rw---- 1 root microk8s 2.2M May 25 07:52 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674966-0000000011675053
1.6M -rw-rw---- 1 root microk8s 1.6M May 25 07:55 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675054-0000000011675115
780K -rw-rw---- 1 root microk8s 780K May 25 07:57 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675116-0000000011675141
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 08:04 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675142-0000000011675462
344K -rw-rw---- 1 root microk8s 342K May 25 08:05 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675463-0000000011675473
3.6M -rw-rw---- 1 root microk8s 3.6M May 25 08:10 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675474-0000000011675621
4.0M -rw-rw---- 1 root microk8s 4.0M May 25 08:15 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675622-0000000011675786
7.3M -rw-rw---- 1 root microk8s 7.3M May 25 08:21 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675787-0000000011676079
4.3M -rw-rw---- 1 root microk8s 4.3M May 25 08:27 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676080-0000000011676249
5.1M -rw-rw---- 1 root microk8s 5.1M May 25 08:31 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676250-0000000011676441
6.0M -rw-rw---- 1 root microk8s 6.0M May 25 08:37 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676442-0000000011676677
4.2M -rw-rw---- 1 root microk8s 4.2M May 25 08:43 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676678-0000000011676852
5.4M -rw-rw---- 1 root microk8s 5.4M May 25 08:49 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676853-0000000011677064
4.4M -rw-rw---- 1 root microk8s 4.4M May 25 08:56 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677065-0000000011677249
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 09:05 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677250-0000000011677583
1.2M -rw-rw---- 1 root microk8s 1.1M May 25 09:06 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677584-0000000011677629
4.3M -rw-rw---- 1 root microk8s 4.3M May 25 09:11 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677630-0000000011677801
6.9M -rw-rw---- 1 root microk8s 6.9M May 25 09:21 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677802-0000000011678086
4.1M -rw-rw---- 1 root microk8s 4.1M May 25 09:27 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678087-0000000011678245
2.1M -rw-rw---- 1 root microk8s 2.1M May 25 09:31 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678246-0000000011678326
2.8M -rw-rw---- 1 root microk8s 2.8M May 25 09:37 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678327-0000000011678443
3.0M -rw-rw---- 1 root microk8s 3.0M May 25 09:42 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678444-0000000011678571
6.6M -rw-rw---- 1 root microk8s 6.6M May 25 09:50 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678572-0000000011678831
3.0M -rw-rw---- 1 root microk8s 3.0M May 25 09:54 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678832-0000000011678949
3.3M -rw-rw---- 1 root microk8s 3.3M May 25 09:59 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678950-0000000011679084
4.2M -rw-rw---- 1 root microk8s 4.2M May 25 10:04 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679085-0000000011679245
1.9M -rw-rw---- 1 root microk8s 1.8M May 25 10:05 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679246-0000000011679315
4.4M -rw-rw---- 1 root microk8s 4.4M May 25 10:12 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679316-0000000011679490
3.9M -rw-rw---- 1 root microk8s 3.9M May 25 10:17 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679491-0000000011679646
4.8M -rw-rw---- 1 root microk8s 4.8M May 25 10:22 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679647-0000000011679841
5.9M -rw-rw---- 1 root microk8s 5.9M May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679842-0000000011680080
4.0K -rw-rw---- 1 root microk8s 1.9K May 13 13:25 /var/snap/microk8s/current/var/kubernetes/backend/cluster.crt
4.0K -rw-rw---- 1 root microk8s 3.2K May 13 13:25 /var/snap/microk8s/current/var/kubernetes/backend/cluster.key
4.0K -rw-rw---- 1 root microk8s  208 May 25 10:30 /var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml
4.0K -rw-rw-r-- 1 root microk8s    2 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/failure-domain
4.0K -rw-rw---- 1 root microk8s   63 May 24 19:00 /var/snap/microk8s/current/var/kubernetes/backend/info.yaml
   0 srw-rw---- 1 root microk8s    0 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/kine.sock
4.0K -rw-rw---- 1 root microk8s   69 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/localnode.yaml
4.0K -rw-rw---- 1 root microk8s   32 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/metadata1
4.0K -rw-rw---- 1 root microk8s   32 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/metadata2
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 10:30 /var/snap/microk8s/current/var/kubernetes/backend/open-1
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/open-2
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/open-3
215M -rw-rw---- 1 root microk8s 215M May 25 09:52 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-451-11678894-46572435
4.0K -rw-rw---- 1 root microk8s  136 May 25 09:52 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-451-11678894-46572435.meta
219M -rw-rw---- 1 root microk8s 219M May 25 10:24 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-464-11679918-48493042
4.0K -rw-rw---- 1 root microk8s  136 May 25 10:24 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-464-11679918-48493042.meta
```

k8s05:
```
1.5M -rw-rw---- 1 root microk8s 1.5M May 25 06:18 /var/snap/microk8s/current/var/kubernetes/backend/0000000011671634-0000000011671693
3.9M -rw-rw---- 1 root microk8s 3.9M May 25 06:22 /var/snap/microk8s/current/var/kubernetes/backend/0000000011671694-0000000011671839
5.0M -rw-rw---- 1 root microk8s 5.0M May 25 06:29 /var/snap/microk8s/current/var/kubernetes/backend/0000000011671840-0000000011672037
4.4M -rw-rw---- 1 root microk8s 4.4M May 25 06:33 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672038-0000000011672212
6.4M -rw-rw---- 1 root microk8s 6.4M May 25 06:39 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672213-0000000011672459
5.0M -rw-rw---- 1 root microk8s 5.0M May 25 06:45 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672460-0000000011672656
4.7M -rw-rw---- 1 root microk8s 4.7M May 25 06:49 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672657-0000000011672845
6.1M -rw-rw---- 1 root microk8s 6.1M May 25 06:55 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672846-0000000011673082
5.9M -rw-rw---- 1 root microk8s 5.9M May 25 07:01 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673083-0000000011673312
5.8M -rw-rw---- 1 root microk8s 5.8M May 25 07:06 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673313-0000000011673541
1.3M -rw-rw---- 1 root microk8s 1.3M May 25 07:08 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673542-0000000011673586
6.9M -rw-rw---- 1 root microk8s 6.9M May 25 07:15 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673587-0000000011673861
7.0M -rw-rw---- 1 root microk8s 7.0M May 25 07:23 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673862-0000000011674145
7.2M -rw-rw---- 1 root microk8s 7.2M May 25 07:30 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674146-0000000011674432
3.5M -rw-rw---- 1 root microk8s 3.5M May 25 07:35 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674433-0000000011674565
7.7M -rw-rw---- 1 root microk8s 7.7M May 25 07:44 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674566-0000000011674881
2.5M -rw-rw---- 1 root microk8s 2.5M May 25 07:49 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674882-0000000011674983
1.8M -rw-rw---- 1 root microk8s 1.8M May 25 07:53 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674984-0000000011675053
1.6M -rw-rw---- 1 root microk8s 1.6M May 25 07:55 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675054-0000000011675114
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 08:03 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675115-0000000011675431
636K -rw-rw---- 1 root microk8s 635K May 25 08:04 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675432-0000000011675456
5.9M -rw-rw---- 1 root microk8s 5.8M May 25 08:12 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675457-0000000011675695
6.9M -rw-rw---- 1 root microk8s 6.9M May 25 08:18 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675696-0000000011675980
5.2M -rw-rw---- 1 root microk8s 5.2M May 25 08:24 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675981-0000000011676182
1.7M -rw-rw---- 1 root microk8s 1.7M May 25 08:27 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676183-0000000011676249
2.1M -rw-rw---- 1 root microk8s 2.1M May 25 08:29 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676250-0000000011676330
7.4M -rw-rw---- 1 root microk8s 7.4M May 25 08:35 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676331-0000000011676614
3.4M -rw-rw---- 1 root microk8s 3.4M May 25 08:40 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676615-0000000011676749
1.2M -rw-rw---- 1 root microk8s 1.2M May 25 08:41 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676750-0000000011676798
5.3M -rw-rw---- 1 root microk8s 5.3M May 25 08:47 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676799-0000000011677006
3.8M -rw-rw---- 1 root microk8s 3.8M May 25 08:52 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677007-0000000011677163
3.7M -rw-rw---- 1 root microk8s 3.7M May 25 08:58 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677164-0000000011677320
5.4M -rw-rw---- 1 root microk8s 5.4M May 25 09:06 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677321-0000000011677547
5.7M -rw-rw---- 1 root microk8s 5.7M May 25 09:10 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677548-0000000011677774
3.7M -rw-rw---- 1 root microk8s 3.7M May 25 09:15 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677775-0000000011677925
3.1M -rw-rw---- 1 root microk8s 3.1M May 25 09:20 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677926-0000000011678055
3.9M -rw-rw---- 1 root microk8s 3.9M May 25 09:25 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678056-0000000011678208
3.0M -rw-rw---- 1 root microk8s 3.0M May 25 09:32 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678209-0000000011678326
400K -rw-rw---- 1 root microk8s 398K May 25 09:33 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678327-0000000011678343
3.6M -rw-rw---- 1 root microk8s 3.6M May 25 09:39 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678344-0000000011678494
3.3M -rw-rw---- 1 root microk8s 3.3M May 25 09:44 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678495-0000000011678627
4.4M -rw-rw---- 1 root microk8s 4.4M May 25 09:49 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678628-0000000011678792
3.8M -rw-rw---- 1 root microk8s 3.8M May 25 09:54 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678793-0000000011678949
2.2M -rw-rw---- 1 root microk8s 2.2M May 25 09:56 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678950-0000000011679037
1.1M -rw-rw---- 1 root microk8s 1.1M May 25 09:58 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679038-0000000011679085
1.8M -rw-rw---- 1 root microk8s 1.8M May 25 10:01 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679086-0000000011679152
4.2M -rw-rw---- 1 root microk8s 4.2M May 25 10:05 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679153-0000000011679314
3.7M -rw-rw---- 1 root microk8s 3.7M May 25 10:10 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679315-0000000011679458
3.7M -rw-rw---- 1 root microk8s 3.7M May 25 10:15 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679459-0000000011679613
940K -rw-rw---- 1 root microk8s 937K May 25 10:17 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679614-0000000011679646
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 10:25 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679647-0000000011679971
1.2M -rw-rw---- 1 root microk8s 1.2M May 25 10:26 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679972-0000000011680017
4.0K -rw-rw---- 1 root microk8s 1.9K May 13 13:25 /var/snap/microk8s/current/var/kubernetes/backend/cluster.crt
4.0K -rw-rw---- 1 root microk8s 3.2K May 13 13:25 /var/snap/microk8s/current/var/kubernetes/backend/cluster.key
4.0K -rw-rw---- 1 root microk8s  208 May 25 10:30 /var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml
4.0K -rw-rw-r-- 1 root microk8s    2 May 25 10:26 /var/snap/microk8s/current/var/kubernetes/backend/failure-domain
4.0K -rw-rw---- 1 root microk8s   64 May 24 18:59 /var/snap/microk8s/current/var/kubernetes/backend/info.yaml
   0 srw-rw---- 1 root microk8s    0 May 25 10:26 /var/snap/microk8s/current/var/kubernetes/backend/kine.sock
4.0K -rw-rw---- 1 root microk8s   70 May 25 10:26 /var/snap/microk8s/current/var/kubernetes/backend/localnode.yaml
4.0K -rw-rw---- 1 root microk8s   32 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/metadata1
4.0K -rw-rw---- 1 root microk8s   32 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/metadata2
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 10:30 /var/snap/microk8s/current/var/kubernetes/backend/open-1
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 10:26 /var/snap/microk8s/current/var/kubernetes/backend/open-2
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 10:26 /var/snap/microk8s/current/var/kubernetes/backend/open-3
213M -rw-rw---- 1 root microk8s 213M May 25 09:49 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-451-11678806-46482303
4.0K -rw-rw---- 1 root microk8s  136 May 25 09:49 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-451-11678806-46482303.meta
217M -rw-rw---- 1 root microk8s 217M May 25 10:22 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-464-11679830-48428239
4.0K -rw-rw---- 1 root microk8s  136 May 25 10:22 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-464-11679830-48428239.meta
```

k8s06:
```
1.5M -rw-rw---- 1 root microk8s 1.5M May 25 06:18 /var/snap/microk8s/current/var/kubernetes/backend/0000000011671634-0000000011671693
3.0M -rw-rw---- 1 root microk8s 3.0M May 25 06:21 /var/snap/microk8s/current/var/kubernetes/backend/0000000011671694-0000000011671802
3.3M -rw-rw---- 1 root microk8s 3.3M May 25 06:26 /var/snap/microk8s/current/var/kubernetes/backend/0000000011671803-0000000011671930
4.4M -rw-rw---- 1 root microk8s 4.4M May 25 06:31 /var/snap/microk8s/current/var/kubernetes/backend/0000000011671931-0000000011672107
2.7M -rw-rw---- 1 root microk8s 2.7M May 25 06:33 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672108-0000000011672212
3.7M -rw-rw---- 1 root microk8s 3.7M May 25 06:37 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672213-0000000011672354
5.0M -rw-rw---- 1 root microk8s 5.0M May 25 06:42 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672355-0000000011672548
6.5M -rw-rw---- 1 root microk8s 6.5M May 25 06:48 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672549-0000000011672807
3.9M -rw-rw---- 1 root microk8s 3.9M May 25 06:52 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672808-0000000011672958
4.3M -rw-rw---- 1 root microk8s 4.3M May 25 06:56 /var/snap/microk8s/current/var/kubernetes/backend/0000000011672959-0000000011673127
6.7M -rw-rw---- 1 root microk8s 6.7M May 25 07:05 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673128-0000000011673393
5.0M -rw-rw---- 1 root microk8s 5.0M May 25 07:08 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673394-0000000011673585
4.2M -rw-rw---- 1 root microk8s 4.2M May 25 07:13 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673586-0000000011673750
3.6M -rw-rw---- 1 root microk8s 3.6M May 25 07:17 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673751-0000000011673900
3.3M -rw-rw---- 1 root microk8s 3.3M May 25 07:21 /var/snap/microk8s/current/var/kubernetes/backend/0000000011673901-0000000011674032
4.1M -rw-rw---- 1 root microk8s 4.1M May 25 07:25 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674033-0000000011674202
3.5M -rw-rw---- 1 root microk8s 3.5M May 25 07:28 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674203-0000000011674342
4.8M -rw-rw---- 1 root microk8s 4.8M May 25 07:33 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674343-0000000011674527
2.3M -rw-rw---- 1 root microk8s 2.3M May 25 07:37 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674528-0000000011674614
3.5M -rw-rw---- 1 root microk8s 3.5M May 25 07:41 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674615-0000000011674758
3.1M -rw-rw---- 1 root microk8s 3.1M May 25 07:44 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674759-0000000011674883
2.0M -rw-rw---- 1 root microk8s 2.0M May 25 07:48 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674884-0000000011674964
2.2M -rw-rw---- 1 root microk8s 2.2M May 25 07:52 /var/snap/microk8s/current/var/kubernetes/backend/0000000011674965-0000000011675053
1.6M -rw-rw---- 1 root microk8s 1.6M May 25 07:55 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675054-0000000011675115
780K -rw-rw---- 1 root microk8s 780K May 25 07:57 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675116-0000000011675141
6.8M -rw-rw---- 1 root microk8s 6.8M May 25 08:03 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675142-0000000011675410
1.7M -rw-rw---- 1 root microk8s 1.7M May 25 08:05 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675411-0000000011675475
2.5M -rw-rw---- 1 root microk8s 2.5M May 25 08:08 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675476-0000000011675578
4.1M -rw-rw---- 1 root microk8s 4.1M May 25 08:13 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675579-0000000011675749
4.0M -rw-rw---- 1 root microk8s 4.0M May 25 08:17 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675750-0000000011675909
5.8M -rw-rw---- 1 root microk8s 5.8M May 25 08:22 /var/snap/microk8s/current/var/kubernetes/backend/0000000011675910-0000000011676144
2.7M -rw-rw---- 1 root microk8s 2.7M May 25 08:27 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676145-0000000011676249
7.3M -rw-rw---- 1 root microk8s 7.3M May 25 08:32 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676250-0000000011676523
3.7M -rw-rw---- 1 root microk8s 3.7M May 25 08:37 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676524-0000000011676677
3.0M -rw-rw---- 1 root microk8s 3.0M May 25 08:41 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676678-0000000011676797
3.7M -rw-rw---- 1 root microk8s 3.7M May 25 08:45 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676798-0000000011676942
3.4M -rw-rw---- 1 root microk8s 3.4M May 25 08:50 /var/snap/microk8s/current/var/kubernetes/backend/0000000011676943-0000000011677084
3.1M -rw-rw---- 1 root microk8s 3.1M May 25 08:54 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677085-0000000011677212
3.4M -rw-rw---- 1 root microk8s 3.4M May 25 08:59 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677213-0000000011677354
3.3M -rw-rw---- 1 root microk8s 3.3M May 25 09:02 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677355-0000000011677491
5.5M -rw-rw---- 1 root microk8s 5.5M May 25 09:08 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677492-0000000011677711
1.7M -rw-rw---- 1 root microk8s 1.7M May 25 09:10 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677712-0000000011677777
2.8M -rw-rw---- 1 root microk8s 2.8M May 25 09:13 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677778-0000000011677893
2.6M -rw-rw---- 1 root microk8s 2.6M May 25 09:18 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677894-0000000011677998
3.0M -rw-rw---- 1 root microk8s 3.0M May 25 09:23 /var/snap/microk8s/current/var/kubernetes/backend/0000000011677999-0000000011678122
2.3M -rw-rw---- 1 root microk8s 2.3M May 25 09:25 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678123-0000000011678210
960K -rw-rw---- 1 root microk8s 957K May 25 09:27 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678211-0000000011678245
2.1M -rw-rw---- 1 root microk8s 2.1M May 25 09:31 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678246-0000000011678326
2.2M -rw-rw---- 1 root microk8s 2.2M May 25 09:36 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678327-0000000011678413
3.0M -rw-rw---- 1 root microk8s 3.0M May 25 09:41 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678414-0000000011678543
676K -rw-rw---- 1 root microk8s 676K May 25 09:42 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678544-0000000011678571
1.6M -rw-rw---- 1 root microk8s 1.6M May 25 09:44 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678572-0000000011678631
3.0M -rw-rw---- 1 root microk8s 3.0M May 25 09:47 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678632-0000000011678744
4.4M -rw-rw---- 1 root microk8s 4.4M May 25 09:52 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678745-0000000011678919
2.9M -rw-rw---- 1 root microk8s 2.9M May 25 09:56 /var/snap/microk8s/current/var/kubernetes/backend/0000000011678920-0000000011679035
2.8M -rw-rw---- 1 root microk8s 2.8M May 25 10:01 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679036-0000000011679149
5.2M -rw-rw---- 1 root microk8s 5.2M May 25 10:07 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679150-0000000011679350
4.4M -rw-rw---- 1 root microk8s 4.4M May 25 10:13 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679351-0000000011679528
5.1M -rw-rw---- 1 root microk8s 5.1M May 25 10:20 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679529-0000000011679736
5.0M -rw-rw---- 1 root microk8s 5.0M May 25 10:24 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679737-0000000011679932
2.9M -rw-rw---- 1 root microk8s 2.9M May 25 10:28 /var/snap/microk8s/current/var/kubernetes/backend/0000000011679933-0000000011680047
812K -rw-rw---- 1 root microk8s 812K May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/0000000011680048-0000000011680080
4.0K -rw-rw---- 1 root microk8s 1.9K May 13 13:25 /var/snap/microk8s/current/var/kubernetes/backend/cluster.crt
4.0K -rw-rw---- 1 root microk8s 3.2K May 13 13:25 /var/snap/microk8s/current/var/kubernetes/backend/cluster.key
4.0K -rw-rw---- 1 root microk8s  208 May 25 10:30 /var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml
4.0K -rw-rw-r-- 1 root microk8s    2 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/failure-domain
4.0K -rw-rw---- 1 root microk8s   63 May 24 17:57 /var/snap/microk8s/current/var/kubernetes/backend/info.yaml
   0 srw-rw---- 1 root microk8s    0 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/kine.sock
4.0K -rw-rw---- 1 root microk8s   69 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/localnode.yaml
4.0K -rw-rw---- 1 root microk8s   32 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/metadata1
4.0K -rw-rw---- 1 root microk8s   32 May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/metadata2
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 10:30 /var/snap/microk8s/current/var/kubernetes/backend/open-1
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/open-2
8.0M -rw-rw---- 1 root microk8s 8.0M May 25 10:29 /var/snap/microk8s/current/var/kubernetes/backend/open-3
213M -rw-rw---- 1 root microk8s 213M May 25 09:49 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-451-11678807-46480546
4.0K -rw-rw---- 1 root microk8s  136 May 25 09:49 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-451-11678807-46480546.meta
217M -rw-rw---- 1 root microk8s 217M May 25 10:22 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-464-11679831-48427924
4.0K -rw-rw---- 1 root microk8s  136 May 25 10:22 /var/snap/microk8s/current/var/kubernetes/backend/snapshot-464-11679831-48427924.meta
```

**Comment by MathieuBordere** [comment]
The contents of the folders look fine, I have seen similar behavior on clusters that have very high storage latency.

Can you run `bcc.biolatency` while running `dd if=/dev/zero of=test.bin bs=4M count=200 conv=fdatasync`
! next command will write an outputfile of 32GB.
and `bcc.biolatency` while running `dd if=/dev/zero of=test.bin bs=4M count=8000 conv=fdatasync` and post both results?

If you're running Ubuntu, the bcc snap contains the bcc.biolatency command.

**Comment by MichaelGooden** [comment]
Results as requested:
```
root@k8s04:/tmp# dd if=/dev/zero of=test.bin bs=4M count=200 conv=fdatasync
200+0 records in
200+0 records out
838860800 bytes (839 MB, 800 MiB) copied, 8.71613 s, 96.2 MB/s
```
```
root@k8s04:~# bcc.biolatency
Tracing block device I/O... Hit Ctrl-C to end.
^C
     usecs               : count     distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 0        |                                        |
        64 -> 127        : 0        |                                        |
       128 -> 255        : 0        |                                        |
       256 -> 511        : 0        |                                        |
       512 -> 1023       : 13       |                                        |
      1024 -> 2047       : 55       |                                        |
      2048 -> 4095       : 127      |*                                       |
      4096 -> 8191       : 57       |                                        |
      8192 -> 16383      : 2281     |**********************                  |
     16384 -> 32767      : 4059     |****************************************|
     32768 -> 65535      : 16       |                                        |
```

```
root@k8s04:/tmp# dd if=/dev/zero of=test.bin bs=4M count=8000 conv=fdatasync
8000+0 records in
8000+0 records out
33554432000 bytes (34 GB, 31 GiB) copied, 308.446 s, 109 MB/s
```
```
root@k8s04:~# bcc.biolatency
Tracing block device I/O... Hit Ctrl-C to end.
^C
     usecs               : count     distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 10       |                                        |
        16 -> 31         : 46       |                                        |
        32 -> 63         : 510      |                                        |
        64 -> 127        : 2592     |                                        |
       128 -> 255        : 2111     |                                        |
       256 -> 511        : 958      |                                        |
       512 -> 1023       : 190      |                                        |
      1024 -> 2047       : 565      |                                        |
      2048 -> 4095       : 4427     |*                                       |
      4096 -> 8191       : 10418    |**                                      |
      8192 -> 16383      : 90087    |***********************                 |
     16384 -> 32767      : 150654   |****************************************|
     32768 -> 65535      : 3361     |                                        |
     65536 -> 131071     : 134      |                                        |
    131072 -> 262143     : 400      |                                        |
    262144 -> 524287     : 8        |                                        |
```

**Comment by MathieuBordere** [comment]
Those results don't look particularly bad.

**Comment by MathieuBordere** [comment]
Can you perform a `hexdump -C` on `metadata1` and `metadata2` on all servers please?

**Comment by MichaelGooden** [comment]
Hopefully GH doesn't mangle these files...

[metadata2-k8s06.txt](https://github.com/ubuntu/microk8s/files/6538375/metadata2-k8s06.txt)
[metadata2-k8s05.txt](https://github.com/ubuntu/microk8s/files/6538376/metadata2-k8s05.txt)
[metadata2-k8s04.txt](https://github.com/ubuntu/microk8s/files/6538377/metadata2-k8s04.txt)
[metadata1-k8s06.txt](https://github.com/ubuntu/microk8s/files/6538378/metadata1-k8s06.txt)
[metadata1-k8s05.txt](https://github.com/ubuntu/microk8s/files/6538379/metadata1-k8s05.txt)
[metadata1-k8s04.txt](https://github.com/ubuntu/microk8s/files/6538380/metadata1-k8s04.txt)

**Comment by MathieuBordere** [comment]
I will look for a way to reproduce this to further help you.

**Comment by MichaelGooden** [comment]
Some notes from my observations:

My understanding of the logs is that the entire kubelite process crashes when either the controller manager or the scheduler cannot obtain a leader lock. I would suspect this is what is causing the high number of Raft elections.

As a diagnostic I increased the leader timeouts significantly for both kube-scheduler and kube-controller-manager, however this doesn't appear to have helped, the same symptoms occur:
```
--leader-elect-lease-duration=60s
--leader-elect-renew-deadline=40s
```

**Comment by balchua** [comment]
I have seen in the logs these types of errors.

```
May 24 18:47:17 k8s05 microk8s.daemon-kubelite[13528]: time="2021-05-24T18:47:17+02:00" level=error msg="error while range on /registry/health : query (try: 0): context deadline exceeded"
May 24 18:47:47 k8s05 microk8s.daemon-kubelite[13528]: time="2021-05-24T18:47:47+02:00" level=error msg="error while range on /registry/cert-manager.io/certificates/ /registry/cert-manager.io/certificates/: query (try: 0): call query-sql (budget 0s): receive: header: EOF"
. . . .
May 24 18:47:47 k8s05 microk8s.daemon-kubelite[13528]: time="2021-05-24T18:47:47+02:00" level=error msg="error while range on /registry/cert-manager.io/certificates/ /registry/cert-manager.io/certificates/: query (try: 0): call query-sql (budget 0s): receive: header: EOF"
. . .
May 24 18:47:53 k8s05 microk8s.daemon-kubelite[13528]: time="2021-05-24T18:47:53+02:00" level=error msg="error in txn: exec (try: 2): call exec-sql (budget 0s): receive: header: EOF"
. . .
```
And then eventually this.

```
May 24 20:22:46 k8s05 microk8s.daemon-kubelite[22269]: I0524 20:22:46.398557   22269 log.go:184] [ERROR] dqlite: proxy: first: remote -> local: read tcp 200.124.153.242:19001->200.124.153.249:50620: read: connection reset by peer
```

**Comment by andrew-landsverk-win** [comment]
I am also seeing these issues since switch from 1.20/stable to 1.21/stable, I have not been able to pinpoint anything further.

**Comment by mikapp** [comment]
I am in the same boat here, can't get 1.21 HA to operate steadily in the same on-prem environment as 1.20 HA. Ended up removing 1.21 and installing it from the 1.20 snap channel. Most of the time it would be pods stuck in the "Terminating" state indefinitely while nodes are "Ready". I have inspection tarballs and can create a separate issue if needed. 
Keep watching this thread for now and thanks for the great product!

**Comment by MichaelGooden** [comment]
@MathieuBordere I have left the nodes as is, moved the workload to a non-HA microk8s cluster for now. I would be happy to run any tests/debug builds/etc against the environment if it can assist.

**Comment by MathieuBordere** [comment]
@MichaelGooden Thanks for this, please keep it in the bad state. I'll align with @ktsakalozos to see what we can provide you to assist with debugging.
